### ISSUE TITLE
normaliseHeaders function looks like it has a case insensitivity bug

### ISSUE THREAD

**Issue by mgsloan** [brendanhay/amazonka#390]
I was taking a look around the code to see how things work and saw this definition for [`normaliseHeaders`](https://github.com/brendanhay/amazonka/blob/c40a49cf08730b746cdb1e162b92a1a0edac7953/core/src/Network/AWS/Sign/V4/Base.hs#L266):

```haskell
normaliseHeaders :: [Header] -> NormalisedHeaders
normaliseHeaders = Tag
    . map    (first CI.foldedCase)
    . nubBy  ((==)    `on` fst)
    . sortBy (compare `on` fst)
    . filter ((/= "authorization") . fst)
```

I didn't encounter any actual issues here, but I think the case folding should be done before everything else.  Otherwise, for example, a header named `Authorization` will not get filtered.

Another thing to consider is that `nubBy` is O(n^2), and it's already done the work of sorting the headers.  May as well use `groupBy` and just take the heads of the groups.  Would probably be fastest to have a `sortNubBy` function, heh.

**Comment by mgsloan** [comment]
I guess it's fine because it's using the CI type (case insensitive).  NVM!
